### DISCUSSION TITLE
URQL graphql-codegen should support `@defer`

### DISCUSSION THREAD

**Discussion by jtwigg** [urql-graphql/urql#3002] [Ideas]
Hope this is the right place, but for `@graphql-codegen/typescript-urql` 
there should be some support for `@defer` because it can  break the typescript contract.

Example:
```
type Song {
  firstVerse: String!
  secondVerse: String!
}
```

```
query SongQuery{
    song {
      firstVerse
      ... @defer {
        secondVerse
      }
    }
}
```

generates:
```
export type SongQueryQuery = {
  __typename: "Query"
    song?:
      | { __typename: "Song"; secondVerse: string; firstVerse: string }
      | undefined
}
```

HOWEVER
because `secondVerse` is deferred, its `undefined` in the first response, and present in the second response. 
```
{
  "firstVerse": "Now I know my ABCs",
  "__typename": "Song"
}
```
followed by 
```
{
  "firstVerse": "Now I know my ABCs",
  "__typename": "Song",
  "secondVerse": "Next time won't you sing with me"
}
```

This means I have to be manually defensive for all of the deferred items in my query.

Possible solutions are
1) Force it to be `undefined`
```
{ __typename: "Song"; secondVerse: string | undefined; firstVerse: string }
```

2) It could possibly be special named `__deferred` 
```
{ __typename: "Song"; secondVerse: string | "__deferred"; firstVerse: string }
```

3) A poorly explored idea of a Promise response could work for PromiseLike responses.. but it might not play well with react hooks.
```
{ __typename: "Song"; secondVerse: Promise<string> | string ; firstVerse: string }
```

**Comment by JoviDeCroock** [comment]
The GraphQL code generator repo is probably a better place to inquire about this. Do note that defer/stream is still not stable so it might not be a priority

**Comment by jtwigg** [comment]
I did find something on this.

https://github.com/dotansimha/graphql-code-generator/issues/7885

Also, the API has been hardened substantially, 
https://github.com/urql-graphql/urql/issues/2995#issuecomment-1455273101
